File: abstractions/typescript/src/serialization/serializationWriterFactoryRegistry.ts
import { SerializationWriter } from "./serializationWriter";
import { SerializationWriterFactory } from "./serializationWriterFactory";

/** This factory holds a list of all the registered factories for the various types of nodes. */
export class SerializationWriterFactoryRegistry implements SerializationWriterFactory {
    /** Default singleton instance of the registry to be used when registring new factories that should be available by default. */
    public static readonly defaultInstance = new SerializationWriterFactoryRegistry();
    public getValidContentType(): string {
        throw new Error("The registry supports multiple content types. Get the registered factory instead.");
    }
    /** List of factories that are registered by content type. */
    public contentTypeAssociatedFactories = new Map<string, SerializationWriterFactory>();
    public getSerializationWriter(contentType: string): SerializationWriter {
        if(!contentType) {
            throw new Error("content type cannot be undefined or empty");
        }
        const factory = this.contentTypeAssociatedFactories.get(contentType);
        if(factory) {
            return factory.getSerializationWriter(contentType);
        } else {
            throw new Error(`Content type ${contentType} does not have a factory registered to be serialized`);
        }
    }

}

File: abstractions/go/serialization/serialization_writer_factory_registry.go
package serialization

import "errors"

// SerializationWriterFactoryRegistry is a factory holds a list of all the registered factories for the various types of nodes.
type SerializationWriterFactoryRegistry struct {
	// ContentTypeAssociatedFactories list of factories that are registered by content type.
	ContentTypeAssociatedFactories map[string]SerializationWriterFactory
}

// DefaultSerializationWriterFactoryInstance is the default singleton instance of the registry to be used when registering new factories that should be available by default.
var DefaultSerializationWriterFactoryInstance = &SerializationWriterFactoryRegistry{
	ContentTypeAssociatedFactories: make(map[string]SerializationWriterFactory),
}

// GetValidContentType returns the valid content type for the SerializationWriterFactoryRegistry
func (m *SerializationWriterFactoryRegistry) GetValidContentType() (string, error) {
	return "", errors.New("the registry supports multiple content types. Get the registered factory instead")
}

// GetSerializationWriter returns the relevant SerializationWriter instance for the given content type
func (m *SerializationWriterFactoryRegistry) GetSerializationWriter(contentType string) (SerializationWriter, error) {
	if contentType == "" {
		return nil, errors.New("the content type is empty")
	}
	factory := m.ContentTypeAssociatedFactories[contentType]
	if factory == nil {
		return nil, errors.New("Content type " + contentType + " does not have a factory registered to be parsed")
	} else {
		return factory.GetSerializationWriter(contentType)
	}
}
